File: validations.py
"""Performs input validations.

Validates on_demand_config.csv, reserves_config.csv, savings_plan_config.csv and total_demand.csv.
Those files are imported by build_simulation.py, which calls the functions on this module.
"""

def validate_on_demand_config(on_demand_config):
    validate_columns('on_demand_config', on_demand_config, ['instance', 'hourly_price'])
    validate_on_demand_instances(on_demand_config)

def validate_columns(file_name, data_frame, names):
    if list(data_frame.columns) != names:
        raise ValueError('The column names in ' + file_name +  ' are incorrect.')

def validate_on_demand_instances(on_demand_config):
    #Checks if the instances are unique
    instances = on_demand_config['instance'].values.tolist()
    instances.sort()
    unique_instances = list(on_demand_config['instance'].value_counts().index)
    unique_instances.sort()

    if (instances != unique_instances):
        raise ValueError('The instances names in on_demand_config should be unique')

def validate_reserves_config(reserves_config, instances):
    validate_columns('reserves_config', reserves_config, ['instance', 'market_name', 'hourly_price', 'upfront_price', 'duration'])
    validate_instances_names('reserves_config', reserves_config, instances)
    validate_reserves_markets(reserves_config, instances)

def validate_instances_names(file_name, data_frame, instances):
    #Checks if the instance names in the data_frame are the same as in the other files
    file_instances = list(data_frame['instance'].value_counts().index)
    file_instances.sort()

    if instances != file_instances:
        raise ValueError('The instances names in on_demand_config and ' + file_name + ' are not the same.')

def validate_reserves_markets(reserves_config, instances):
    #All instances should have the same reserve markets
    instance_line = reserves_config[reserves_config['instance'] == instances[0]]
    markets = list(instance_line.loc[:,'market_name'])
    markets.sort()
    previous_markets = markets
    
    for instance in instances:
        instance_line = reserves_config[reserves_config['instance'] == instance]
        markets = list(instance_line.loc[:,'market_name'])
        markets.sort()
        if markets != previous_markets:
            raise ValueError('Not all instances have the same reserve market names.')
        previous_markets = markets

def validate_savings_plan_config(savings_plan_config, instances):
    validate_columns('savings_plan_config', savings_plan_config, ['instance', 'hourly_price', 'duration'])
    validate_instances_names('savings_plan_config', savings_plan_config, instances)
    validate_savings_plan_durations(savings_plan_config)

def validate_savings_plan_durations(savings_plan_config):
    #All savings plan durations should be the same
    savings_plan_durations = list(savings_plan_config['duration'].value_counts().index)
    if len(savings_plan_durations) != 1:
        raise ValueError('All instances must have the same savings plan duration.')

def validate_demand(raw_demand, instances):
    if raw_demand.columns[0] != 'hour': #could i just correct in the code?
        raise ValueError('The first column name in the demand file is incorrect.')
    
    validate_demand_instances(raw_demand, instances)

def validate_demand_instances(raw_demand, instances):
    #the demand should contain all the instances passed on the input
    demand_col = list(raw_demand.columns)
    for instance in instances:
        if instance not in demand_col:
            raise ValueError('The instance ' + instance + ' is not on the demand file.')

File: cpp/validations.cpp
#include "csv_parser.h"
#include <algorithm>
#include <iostream>
#include <set>
#include <string>
#include <vector>

using namespace std;

void validate_columns(vector<vector<string>> dataframe, vector<string> names,
                      string file_name) {
  if (dataframe[0] != names) {
    cerr << "The column names in " << file_name << " are incorrect.";
  }
}

void validate_on_demand_instances(vector<vector<string>> on_demand_config) {
  vector<string> instances = get_column(on_demand_config, "instance");
  set<string> unique_strings;

  for (string inst : instances) {
    if (!unique_strings.insert(inst).second) {
      cerr << "The instances names in on_demand_config should be unique";
    }
  }
}

void validate_on_demand_config(vector<vector<string>> on_demand_config) {
  validate_columns(on_demand_config, {"instance", "hourly_price"},
                   "on_demand_config");
  validate_on_demand_instances(on_demand_config);
}

void validate_instances_names(vector<vector<string>> dataframe,
                              vector<string> instances, string file_name) {
  vector<string> file_instances = get_column(dataframe, "instance");

  sort(file_instances.begin(), file_instances.end());
  sort(instances.begin(), instances.end());

  if (file_instances != instances) {
    cerr << "The instance names in on_demand_config and " << file_name
         << " are not the same.";
  }
}

void validate_savings_plan_durations(
    vector<vector<string>> savings_plan_config) {
  vector<string> savings_plan_durations =
      get_column(savings_plan_config, "duration");
  set<string> unique_elements;

  for (string duration : savings_plan_durations) {
    unique_elements.insert(duration);

    if (unique_elements.size() > 1) {
      cerr << "All instances must have the same savings plan duration";
    }
  }
}

void validate_savings_plan_config(vector<vector<string>> savings_plan_config,
                                  vector<string> instances) {
  validate_columns(savings_plan_config,
                   {"instance", "hourly_price", "duration"},
                   "savings_plan_config");
  validate_instances_names(savings_plan_config, instances,
                           "savings_plan_config");
  validate_savings_plan_durations(savings_plan_config);
}

void validate_demand_instances(vector<vector<string>> raw_demand,
                               vector<string> instances) {
  vector<string> demand_col = raw_demand[0];

  for (string instance : instances) {
    if (find(demand_col.begin(), demand_col.end(), instance) ==
        demand_col.end()) {
      cerr << "The instance " << instance << " is not on the demand file.";
    }
  }
}

void validate_demand(vector<vector<string>> raw_demand,
                     vector<string> instances) {
  if (raw_demand[0][0] != "hour") {
    cerr << "The first column name in the demand file is incorrect.";
  }

  validate_demand_instances(raw_demand, instances);
}
